### PR TITLE
KAZOO-5905: Split kz_cache and gen_listener functionality into separate modules

### PR DESCRIPTION
**Pull request by harenson**

### CODE REVIEW

File: core/kazoo_data/src/kzs_cache.erl [hunk 1]
```diff
@@ -262,3 +267,157 @@ flush_cache_docs(Db, Docs, Options) ->
...
+    ,'delivery_tag' = rand:uniform(1000)
+    ,'exchange' = Category
+    ,'routing_key' = RKey
+    }.
+
+-spec load_test_build_pbasic() -> #'P_basic'{}.
+load_test_build_pbasic() ->
+    #'P_basic'
+    {'content_type' = <<"application/json">>
+    ,'content_encoding' = 'undefined'
```
Comment (jamesaimonetti) [inline core/kazoo_data/src/kzs_cache.erl]: Since these are records, the `undefined` value is implied; remove these until such time as they are needed.

Comment (harenson) [reply core/kazoo_data/src/kzs_cache.erl]: Done, thanks 👍

File: core/kazoo_caches/src/kazoo_caches_sup.erl [hunk 2]
```diff
@@ -45,7 +45,7 @@ start_link() ->
 %%------------------------------------------------------------------------------
 -spec init(any()) -> kz_types:sup_init_ret().
 init([]) ->
-    RestartStrategy = 'one_for_one',
+    RestartStrategy = 'one_for_all',
```
Comment (lazedo) [inline core/kazoo_caches/src/kazoo_caches_sup.erl]: why this change here ? if one cache crashes why stop the other ?

File: core/kazoo_caches/src/kz_cache.erl [hunk 3]
```diff
@@ -37,37 +37,29 @@
...
         ,terminate/2
         ,code_change/3
         ]).
 
+%% This functions are meant to be used only by `kz_cache_listener' module.
+-export([init_state/3
+        ,tab/1
+        ,pointer_tab/1
+        ,monitor_tab/1
+        ,erase_changed/2]).
```
Comment (jamesaimonetti) [inline core/kazoo_caches/src/kz_cache.erl]: Drop the `]).` to a newline

File: core/kazoo_caches/include/kazoo_caches.hrl [hunk 4]
```diff
@@ -1,10 +1,12 @@
 -ifndef(KAZOO_CACHES_HRL).
 
+-include_lib("kazoo_stdlib/include/kz_types.hrl").
+
 -define(KAPPS_CONFIG_CACHE, 'kapps_config_cache').
 -define(KAPPS_CALL_CACHE, 'kapps_call_cache').
 -define(KAPPS_GETBY_CACHE, 'kapps_getby_cache').
-
--include_lib("kazoo_stdlib/include/kz_types.hrl").
+-define(KAPPS_LISTENER_CACHE, 'kapps_listener_cache').
```
Comment (jamesaimonetti) [inline core/kazoo_caches/include/kazoo_caches.hrl]: Seems unused?

File: core/kazoo_caches/src/kz_cache.erl [hunk 5]
```diff
@@ -78,11 +70,13 @@
...
                ,expire_node_flush = 'false' :: boolean()
                ,expire_period = ?EXPIRE_PERIOD :: timeout()
-               ,expire_period_ref :: reference()
+               ,expire_period_ref :: reference() | 'undefined'
                ,props = [] :: kz_term:proplist()
                ,has_monitors = 'false' :: boolean()
                }).
+
 -type state() :: #state{}.
+-export_type([store_options/0, state/0]).
```
Comment (jamesaimonetti) [inline core/kazoo_caches/src/kz_cache.erl]: Why would `#state{}` be exported as a type? It is internal to the gen_server, no?

File: core/kazoo_caches/src/kz_cache.erl [hunk 6]
```diff
@@ -357,6 +329,7 @@ init(Name, ExpirePeriod, Props, 'undefined') ->
     init(Name, ExpirePeriod, Props);
 init(Name, ExpirePeriod, Props, _Bindings) ->
     kapi_conf:declare_exchanges(),
+    _ = kz_cache_listener:maybe_start_listener(Name, Props),
```
Comment (jamesaimonetti) [inline core/kazoo_caches/src/kz_cache.erl]: I thought we were starting the listener as a child of a supervisor?

File: core/kazoo_caches/src/kz_cache_listener.erl [hunk 7]
```diff
@@ -0,0 +1,231 @@
+-module(kz_cache_listener).
```
Comment (jamesaimonetti) [inline core/kazoo_caches/src/kz_cache_listener.erl]: Missing edoc module header comment thingy

File: core/kazoo_stdlib/include/kz_types.hrl [hunk 8]
```diff
@@ -55,7 +55,7 @@
 -define(SUPER_NAME_TYPE(N, I, Type), {N, {I, 'start_link', []}, Type, 'infinity', 'supervisor', [I]}).
 
 -define(CACHE(N), {N, {'kz_cache', 'start_link', [N]}, 'permanent', 5 * ?MILLISECONDS_IN_SECOND, 'worker', ['kz_cache']}).
--define(CACHE_ARGS(N, Arg), {N, {'kz_cache', 'start_link', [N, Arg]}, 'permanent', 5 * ?MILLISECONDS_IN_SECOND, 'worker', ['kz_cache']}).
+-define(CACHE_ARGS(N, Arg), {N, {'kz_cache', 'start_link', [N, Arg]}, 'permanent', 5 * ?MILLISECONDS_IN_SECOND, 'worker', ['kz_cache_listener']}).
```
Comment (jamesaimonetti) [inline core/kazoo_stdlib/include/kz_types.hrl]: I don't think this change is necessary, is it?
